### ISSUE TITLE
"MUST return 400 for bad relationship path" causes unnecessary breaking changes

### ISSUE THREAD

**Issue by cmeeren** [json-api/json-api#1459]
The spec says this:

> If a server is unable to identify a relationship path or does not support inclusion of resources from a path, it MUST respond with 400 Bad Request.

Consider an operation `GET /vehicles` that return two types, `truck` and `car`. A `truck` has a relationship `trailers`, and a `car` does not.

Clients send requests to `GET /vehicles?include=trailers` to search for both trucks and cars, with the trailers of any `truck` that is returned as included data.

If the API changes so that `GET /vehicles` no longer return `trucks`, that in itself may not be a breaking change (from the client's perspective, the only change is that no trucks turn up when searching). However, with the clause quoted above, the API must now return 400 for `GET /vehicles?include=trailers`. This seems to be an unnecessary breaking change.

Is this intentional? If so, why?

**Comment by kaitoqueiroz** [comment]
~~You're trying to implement a polymorphic resource, and it is not supported.~~

Please take a look into this discussion: [https://discuss.jsonapi.org/t/relationships-within-abstract-relationship/1158](https://discuss.jsonapi.org/t/relationships-within-abstract-relationship/1158)

**Comment by cmeeren** [comment]
I don't see anything in the discussion you linked to which says that collections (whether top-level or to-many relationships) may not contain multiple resource types. The closest I can find is a warning that you have to be careful with the `include` parameter.

Furthermore, Michael explicitly says in [this short discussion](https://discuss.jsonapi.org/t/endpoints-vs-types/1265) that it is supported.

> **Polymorphism is supported in the spec, and even unrelated json:api types can be in the same collections.** This is why it’s important you don’t try to infer any information from coincidental likeness in deployment and documents.

(Emphasis mine)

**Comment by kaitoqueiroz** [comment]
@cmeeren Sorry, I was referring to this:
> I’m not sure if this is the API you’re working with or a contrived example, but `the problem here looks like you want the cars relationship to be polymorphic, and then utilize a feature of {json:api} which doesn’t support it`.

**Comment by cmeeren** [comment]
Yes, I read that to mean that JSON:API's `include` parameter does not support polymorphic relationships (and it says nothing about whether polymorphic collections/relationships are supported in the first place, but the other source I linked to says it is supported). Is that your interpretation, too?

In any case, the spec itself does not give me any indications that using `include` for polymorphic relationships is not supported, and I don't see a problem with it. My JSON:API framework, [Felicity](https://github.com/cmeeren/Felicity), supports this without problems. I.e. `GET /vehicles?include=trailers` as described in the OP works fine, except for the "MUST return 400" clause I am referring to. There are no technical or logical problems with implementing it (in fact, the way I have implemented relationship traversal in the framework, I would actively have to block polymorphic relationship includes if that was desired), though of course it can get confusing from a user perspective if not used carefully.

**Comment by kaitoqueiroz** [comment]
Yes, my statement that implementing a polymorphic resource was not supported, was a misinterpretation from my side (I edited my comment). 
My interpretation, indeed, was that the `include` feature doesn't support polymorphic relationship. This can be an argument for why they opted to use "MUST return 400".

**Comment by cmeeren** [comment]
I see. Thanks for the clarification.

It would be good to have the thoughts of someone "officially" involved in the JSON:API spec. (I assume you're not since you don't have a member/collaborator/contributor badge here on GH; forgive me if I'm mistaken.)

**Comment by cmeeren** [comment]
Bumping this: Could anyone officially involved in the JSON:API spec say something about this issue/discussion?

**Comment by dgeb** [comment]
Thanks for asking about this and sorry for the delay in responding. This situation is probably underspecified and could use some clarification. Let's look at the condition in question:

> If a server is unable to identify a relationship path or does not support inclusion of resources from a path ...

This condition leaves the decision entirely up to the server. It's up to the server to decide if it can "identify a relationship path" and "support inclusion of resources" from that path. It seems entirely reasonable that a server could choose to include related resources only for types that support a particular relationship path. Returning a 400 in this case serves no one particularly well _unless the server decides that such a relationship path is unsupported_. The "MUST return 400" clause is only present so servers know how to communicate complete lack of support for the requested includes.

**Comment by cmeeren** [comment]
Thanks for the clarification! As I understand it, it is then fine that my JSON:API framework Felicity (mentioned above) supports including related resources from polymorphic endpoints, and never returns 400.

Unknown relationship paths would in most cases indicate a client error, and a 400 would be a helpful response in those cases, but as I described, returning 400 could lead to unnecessary breaking changes because no framework can by itself discern between non-existent relationships where 400 is the obvious correct choice in all cases, and relationships that used to exist but was removed in a backwards-compatible way as I described.

Actually, I can think of non-polymorphic examples where this is the case, too. Consider a resource with a relationship that is only present in the response if the user is authorized to view that relationship for the resource. In other words, the client must handle the case where it is missing, even if `include`d. It would then be a non-breaking change in itself to remove the relationship entirely, but if the server then has to return 400 if the (now non-existent) relationship is `include`d, that would be breaking.

In any case, I think my question here is answered (unless the above contains any misconceptions), so I'm closing. Feel free to clarify any incorrect assumptions or reasoning.
